source/Calamari.Shared/Integration/Packages/Download/DockerImagePackageDownloader.cs

```diff
@@ -122,25 +169,71 @@ void PerformPull(string? username, string? password, string fullImageName, strin
             if (!long.TryParse(parts[1], out var size))
             {
                 size = 0;
-                Log.Verbose($"Unable to parse image size. ({parts[0]})");
+                log.Verbose($"Unable to parse image size. ({parts[0]})");
             }
 
 
             return (hash, size);
         }
 
-        string GetFetchScript()
+        IEnumerable<string>? GetCachedImageDigests()
+        {
+            var output = "";
+            var result = SilentProcessRunner.ExecuteCommand("docker",
+                                                            "image ls --format=\"{{.ID}}\" --no-trunc",
+                                                            ".", 
+                                                            environmentVariables, 
+                                                            (stdout) => { output += stdout + " "; },
+                                                            (error) => { });
+            return result.ExitCode == 0 
+                ? output.Split(' ').Select(digest => digest.Trim()) 
+                : null;
+        }
+
+        IEnumerable<string>? GetImageDigests(string fullImageName)
+        {
+            var output = "";
+            var result = SilentProcessRunner.ExecuteCommand("docker",
+                                                            $"manifest inspect --verbose {fullImageName}",
+                                                            ".", 
+                                                            environmentVariables, 
+                                                            (stdout) => { output += stdout; },
+                                                            (error) => { });
+
+            if (result.ExitCode != 0)
+            {
+                return null;
```

Comment:
Would it helpful to know that something went wrong with trying to get image digests? The end message saying "image may not be cached" may not really need to change depending on whether we were able to query digests successfully or not, so potentially it may not really matter that much to us.

Reply:
Basically I think we would swallow the errors encountered when trying to get the digest information. This is because the aim of this change is to add a friendly information-level log to help users understand what is potentially causing the deploy to run longer than expected. Therefore, if we report the errors encountered in making these checks, that is likely to pollute the logs with irrelevant information to the users. So here if there are any errors happened during checking whether the digest is cached, we simply treat the image as being cached and don't write the log. (But when we really need to do debug something, we can always know if the image is being cached by checking the logs from docker pull).